syntax = "proto3";

package raftRpcProtoc;

option cc_generic_services = true;

message LogEntry{
    bytes Command = 1;
    int32 LogTerm=2;
    int32 LogIndex=3;
}

message AppendEntriesRequest{
    int32 Term=1;
    int32 LeaderId=2;
    int32 PrevLogIndex=3;
    int32 PrevLogTerm=4;
    repeated LogEntry Entries=5;
    int32 LeaderCommit=6;
}
message AppendEntriesResponse{
    int32 term=1;
    bool success=2;
    int32 UpdateNextIndex = 3;//快速调整leader对应的nextIndex
}

message RequestVoteRequest{
    int32 Term=1;
    int32 CandidateId=2;
    int32 LastLogIndex=3;
    int32 LastLogTerm=4;
}
message RequestVoteResponse{
    int32 Term=1;
    bool VoteGranted=2;//是否给我投票了
    int32 VoteState=3;//该节点的投票状态
}
message InstallSnapshotRequest{
    int32 Term=1;
    int32 LeaderId=2;
    int32 LastIncludedIndex=3;
    int32 LastIncludedTerm=4;
    bytes Data=5;
}
message InstallSnapshotResponse{
    int32 Term=1;
}
service raftRpc{
    rpc AppendEntries(AppendEntriesRequest) returns(AppendEntriesResponse);
    rpc RequestVote(RequestVoteRequest)returns(RequestVoteResponse);
    rpc InstallSnapshot(InstallSnapshotRequest)returns(InstallSnapshotResponse);
}